File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/model/SprayTankActual.kt
package com.rork.vinetrack.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Exact shared JSON contract for one confirmed product quantity. */
@Serializable
data class SprayTankActualChemical(
    val id: String,
    val plannedChemicalId: String? = null,
    val savedChemicalId: String? = null,
    val name: String,
    val actualAmountBase: Double,
    val unit: String,
    val replacesPlannedChemicalId: String? = null,
    val usageKind: String? = null,
    val productCategory: String? = null,
    val physicalForm: String? = null,
    val snapshotAt: String? = null,
) {
    init {
        require(id.isNotBlank() && actualAmountBase.isFinite() && actualAmountBase >= 0.0)
        require(unit in setOf("Litres", "mL", "Kg", "g"))
    }
}

@Serializable
data class SprayTankActual(
    val id: String,
    @SerialName("vineyard_id") val vineyardId: String,
    @SerialName("spray_record_id") val sprayRecordId: String,
    @SerialName("trip_id") val tripId: String,
    @SerialName("tank_session_id") val tankSessionId: String,
    @SerialName("tank_number") val tankNumber: Int,
    @SerialName("water_volume_l") val waterVolumeL: Double? = null,
    val chemicals: List<SprayTankActualChemical>,
    @SerialName("confirmed_at") val confirmedAt: String,
    @SerialName("confirmed_by") val confirmedBy: String,
    @SerialName("client_updated_at") val clientUpdatedAt: String = confirmedAt,
    @SerialName("correction_version") val correctionVersion: Long = 0,
    @SerialName("last_corrected_at") val lastCorrectedAt: String? = null,
) {
    init {
        require(tankSessionId.isNotBlank() && tankNumber >= 1)
        require(waterVolumeL == null || (waterVolumeL.isFinite() && waterVolumeL >= 0.0))
    }
}

/** Resolves one exact trip/spray/session actual and rejects ambiguous duplicate revisions. */
fun resolveSprayTankActual(
    plannedTank: SprayTank,
    actuals: List<SprayTankActual>,
    vineyardId: String,
    sprayRecordId: String,
    tripId: String,
    tankSessionIds: Set<String> = emptySet(),
): SprayTankActual? {
    if (tankSessionIds.isEmpty()) return null
    val scoped = actuals.filter { actual ->
        actual.vineyardId == vineyardId && actual.sprayRecordId == sprayRecordId &&
            actual.tripId == tripId && actual.tankNumber == plannedTank.tankNumber &&
            actual.tankSessionId in tankSessionIds
    }
    val bySession = scoped.groupBy { it.tankSessionId }
    if (bySession.size != 1) return null
    val revisions = bySession.values.single()
    val highestVersion = revisions.maxOfOrNull { it.correctionVersion } ?: return null
    val highest = revisions.filter { it.correctionVersion == highestVersion }
    return highest.singleOrNull()
}

/** True only when exact identity, water, and every planned or amended chemical result are unambiguous. */
fun areSprayTankActualsComplete(
    plannedTanks: List<SprayTank>,
    actuals: List<SprayTankActual>,
    vineyardId: String,
    sprayRecordId: String,
    tripId: String,
    tankSessionIdsByNumber: Map<Int, Set<String>> = emptyMap(),
): Boolean {
    if (plannedTanks.isEmpty()) return false
    val plannedNumbers = plannedTanks.mapTo(mutableSetOf()) { it.tankNumber }
    val scopedActuals = actuals.filter {
        it.vineyardId == vineyardId && it.sprayRecordId == sprayRecordId && it.tripId == tripId
    }
    if (scopedActuals.any { actual ->
            actual.tankNumber !in plannedNumbers ||
                actual.tankSessionId !in tankSessionIdsByNumber[actual.tankNumber].orEmpty()
        }) return false
    return plannedTanks.all { tank ->
        val actual = resolveSprayTankActual(
            tank, actuals, vineyardId, sprayRecordId, tripId,
            tankSessionIdsByNumber[tank.tankNumber].orEmpty(),
        ) ?: return@all false
        if (actual.waterVolumeL?.let { it.isFinite() && it >= 0.0 } != true) return@all false
        val plannedIds = tank.chemicals.mapTo(mutableSetOf()) { it.id }
        if (plannedIds.size != tank.chemicals.size || actual.chemicals.map { it.id }.toSet().size != actual.chemicals.size) return@all false
        val validAssociations = actual.chemicals.all { line ->
            if (!line.actualAmountBase.isFinite() || line.actualAmountBase < 0.0) return@all false
            when (line.usageKind ?: "planned") {
                "planned" -> line.plannedChemicalId in plannedIds && line.replacesPlannedChemicalId == null
                "additional" -> line.plannedChemicalId == null && line.replacesPlannedChemicalId == null
                "substitution" -> line.plannedChemicalId == null && line.replacesPlannedChemicalId in plannedIds
                else -> false
            }
        }
        validAssociations && tank.chemicals.all { planned ->
            val direct = actual.chemicals.filter { (it.usageKind ?: "planned") == "planned" && it.plannedChemicalId == planned.id }
            val substitutes = actual.chemicals.filter { it.usageKind == "substitution" && it.replacesPlannedChemicalId == planned.id }
            when {
                substitutes.size == 1 -> direct.size <= 1 && direct.all { it.actualAmountBase == 0.0 }
                substitutes.isEmpty() -> direct.size == 1
                else -> false
            }
        }
    }
}

File: ios/VineTrack/LegacyImported/Models/SprayTankActual.swift
import Foundation

/// Frozen confirmation of what was actually placed in one started spray tank.
nonisolated struct SprayTankActualChemical: Codable, Identifiable, Sendable, Hashable {
    let id: UUID
    let plannedChemicalId: UUID?
    let savedChemicalId: UUID?
    let replacesPlannedChemicalId: UUID?
    let usageKind: String?
    let name: String
    let actualAmountBase: Double
    let unit: ChemicalUnit

    init(
        id: UUID = UUID(),
        plannedChemicalId: UUID?,
        savedChemicalId: UUID?,
        replacesPlannedChemicalId: UUID? = nil,
        usageKind: String? = nil,
        name: String,
        actualAmountBase: Double,
        unit: ChemicalUnit
    ) throws {
        guard actualAmountBase.isFinite, actualAmountBase >= 0 else { throw SprayTankActualValidationError.invalidAmount }
        self.id = id
        self.plannedChemicalId = plannedChemicalId
        self.savedChemicalId = savedChemicalId
        self.replacesPlannedChemicalId = replacesPlannedChemicalId
        self.usageKind = usageKind
        self.name = name
        self.actualAmountBase = actualAmountBase
        self.unit = unit
    }

    var displayAmount: Double { unit.fromBase(actualAmountBase) }
}

nonisolated struct SprayTankActual: Codable, Identifiable, Sendable, Hashable {
    let id: UUID
    let vineyardId: UUID
    let sprayRecordId: UUID
    let tripId: UUID
    let tankSessionId: String
    let tankNumber: Int
    let waterVolumeL: Double?
    let chemicals: [SprayTankActualChemical]
    let confirmedAt: Date
    let confirmedBy: UUID
    let clientUpdatedAt: Date
    let correctionVersion: Int?
    let lastCorrectedAt: Date?

    init(
        id: UUID = UUID(),
        vineyardId: UUID,
        sprayRecordId: UUID,
        tripId: UUID,
        tankSessionId: String,
        tankNumber: Int,
        waterVolumeL: Double?,
        chemicals: [SprayTankActualChemical],
        confirmedAt: Date,
        confirmedBy: UUID,
        clientUpdatedAt: Date? = nil,
        correctionVersion: Int? = nil,
        lastCorrectedAt: Date? = nil
    ) throws {
        guard tankNumber >= 1, !tankSessionId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              waterVolumeL.map({ $0.isFinite && $0 >= 0 }) ?? true,
              chemicals.allSatisfy({ $0.actualAmountBase.isFinite && $0.actualAmountBase >= 0 })
        else { throw SprayTankActualValidationError.invalidAmount }
        self.id = id
        self.vineyardId = vineyardId
        self.sprayRecordId = sprayRecordId
        self.tripId = tripId
        self.tankSessionId = tankSessionId
        self.tankNumber = tankNumber
        self.waterVolumeL = waterVolumeL
        self.chemicals = chemicals
        self.confirmedAt = confirmedAt
        self.confirmedBy = confirmedBy
        self.clientUpdatedAt = clientUpdatedAt ?? confirmedAt
        self.correctionVersion = correctionVersion
        self.lastCorrectedAt = lastCorrectedAt
    }
}

nonisolated func resolveSprayTankActual(
    plannedTank: SprayTank,
    actuals: [SprayTankActual],
    vineyardId: UUID,
    sprayRecordId: UUID,
    tripId: UUID,
    tankSessionIds: Set<String>? = nil
) -> SprayTankActual? {
    if let tankSessionIds, tankSessionIds.isEmpty { return nil }
    let scoped = actuals.filter { actual in
        actual.vineyardId == vineyardId && actual.sprayRecordId == sprayRecordId &&
            actual.tripId == tripId && actual.tankNumber == plannedTank.tankNumber &&
            (tankSessionIds == nil || tankSessionIds!.contains(actual.tankSessionId))
    }
    let bySession = Dictionary(grouping: scoped, by: \.tankSessionId)
    guard bySession.count == 1, let revisions = bySession.values.first,
          let highestVersion = revisions.map({ $0.correctionVersion ?? 0 }).max()
    else { return nil }
    let highest = revisions.filter { ($0.correctionVersion ?? 0) == highestVersion }
    return highest.count == 1 ? highest[0] : nil
}

nonisolated func areSprayTankActualsComplete(
    plannedTanks: [SprayTank],
    actuals: [SprayTankActual],
    vineyardId: UUID,
    sprayRecordId: UUID,
    tripId: UUID,
    tankSessionIdsByNumber: [Int: Set<String>]? = nil
) -> Bool {
    guard !plannedTanks.isEmpty else { return false }
    return plannedTanks.allSatisfy { tank in
        guard let actual = resolveSprayTankActual(
            plannedTank: tank, actuals: actuals, vineyardId: vineyardId,
            sprayRecordId: sprayRecordId, tripId: tripId,
            tankSessionIds: tankSessionIdsByNumber?[tank.tankNumber]
        ), let water = actual.waterVolumeL, water.isFinite, water >= 0
        else { return false }
        let plannedIds = Set(tank.chemicals.map(\.id))
        guard actual.chemicals.allSatisfy({ line in
            guard line.actualAmountBase.isFinite, line.actualAmountBase >= 0 else { return false }
            switch line.usageKind ?? "planned" {
            case "planned": return line.plannedChemicalId.map(plannedIds.contains) == true && line.replacesPlannedChemicalId == nil
            case "additional": return line.plannedChemicalId == nil && line.replacesPlannedChemicalId == nil
            case "substitution": return line.plannedChemicalId == nil && line.replacesPlannedChemicalId.map(plannedIds.contains) == true
            default: return false
            }
        }) else { return false }
        return tank.chemicals.allSatisfy { planned in
            let direct = actual.chemicals.filter { ($0.usageKind ?? "planned") == "planned" && $0.plannedChemicalId == planned.id }.count
            let substitutions = actual.chemicals.filter { $0.usageKind == "substitution" && $0.replacesPlannedChemicalId == planned.id }.count
            return direct + substitutions == 1
        }
    }
}

nonisolated enum SprayTankActualValidationError: LocalizedError, Sendable {
    case invalidAmount
    case unavailablePlan
    case localSaveFailed

    var errorDescription: String? {
        switch self {
        case .invalidAmount: "Enter zero or a positive finite amount for every field."
        case .unavailablePlan: "Planned tank mix unavailable."
        case .localSaveFailed: "The tank was not started because its actual mix could not be saved locally."
        }
    }
}
